substrate/frame/staking/ah-client/src/lib.rs

```diff
@@ -0,0 +1,318 @@
+			ValidatorSet::<T>::put(Some(new_validator_set));
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> historical::SessionManager<T::AccountId, Exposure<T::AccountId, BalanceOf<T>>>
+		for Pallet<T>
+	{
+		fn new_session(
+			_: sp_staking::SessionIndex,
+		) -> Option<Vec<(T::AccountId, Exposure<T::AccountId, BalanceOf<T>>)>> {
+			// If there is a new validator set - return it. Otherwise return `None`.
+			ValidatorSet::<T>::take()
+		}
+
+		fn new_session_genesis(
+			_: SessionIndex,
+		) -> Option<Vec<(T::AccountId, Exposure<T::AccountId, BalanceOf<T>>)>> {
+			ValidatorSet::<T>::take()
+		}
+
+		fn start_session(start_index: SessionIndex) {
+			<Self as pallet_session::SessionManager<_>>::start_session(start_index)
+		}
+
+		fn end_session(end_index: SessionIndex) {
+			<Self as pallet_session::SessionManager<_>>::end_session(end_index)
+		}
+	}
+
+	impl<T: Config> pallet_session::SessionManager<T::AccountId> for Pallet<T> {
+		fn new_session(_: u32) -> Option<Vec<<T as frame_system::Config>::AccountId>> {
+			// Doesn't do anything because all the logic is handled in `historical::SessionManager`
+			// implementation
+			defensive!("new_session should not be called");
+			None
+		}
+
+		fn end_session(session_index: u32) {
+			let authors = BlockAuthors::<T>::iter().collect::<Vec<_>>();
+			// The maximum number of block authors is `num_cores * max_validators_per_core` (both
+			// are parameters from [`SchedulerParams`]).
+			let _ = BlockAuthors::<T>::clear(u32::MAX, None);
+
+			let message = Xcm(vec![
+				Instruction::UnpaidExecution {
+					weight_limit: WeightLimit::Unlimited,
+					check_origin: None,
+				},
+				mk_asset_hub_call::<T>(StakingCalls::EndSession(session_index, authors)),
+			]);
+
+			if let Err(err) = send_xcm::<T::SendXcm>(
+				Location::new(0, [Junction::Parachain(T::AssetHubId::get())]),
+				message,
+			) {
+				log::error!(target: LOG_TARGET, "Sending `EndSession` to AssetHub failed: {:?}", err);
+			}
+		}
+
+		fn start_session(session_index: u32) {
+			// TODO: pass active validator set somehow???
+
+			let message = Xcm(vec![
+				Instruction::UnpaidExecution {
+					weight_limit: WeightLimit::Unlimited,
+					check_origin: None,
+				},
+				mk_asset_hub_call::<T>(StakingCalls::StartSession(session_index)),
+			]);
+			if let Err(err) = send_xcm::<T::SendXcm>(
+				Location::new(0, [Junction::Parachain(T::AssetHubId::get())]),
+				message,
+			) {
+				log::error!(target: LOG_TARGET, "Sending `StartSession` to AssetHub failed: {:?}", err);
+			}
+		}
+	}
+
+	impl<T> pallet_authorship::EventHandler<T::AccountId, BlockNumberFor<T>> for Pallet<T>
+	where
+		T: Config + pallet_authorship::Config + pallet_session::Config + crate::Config,
+		T::AccountId: Into<AccountId32>,
+	{
+		// Notes the authored block in `BlockAuthors`.
+		fn note_author(author: T::AccountId) {
+			BlockAuthors::<T>::mutate(author.into(), |block_count| {
```

Comment:
probably better to call it session points or something and keep track of that. It will be similar to `ErasRewardPoints`, and validator receives these for both block authoring as well as [backing para-blocks](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/runtime/parachains/src/reward_points.rs#L64).